### PR TITLE
Replace Mock User IDs with Real User IDs and Implement New User Profile Management

### LINKED ISSUE #289: Replace Mock User ID with Real ID and Implement User Creation Service
**Related User Story**
#120 

**Description**
This ticket involves replacing the existing mock user ID with a real user ID both on the frontend and backend. Additionally, a user creation service should be implemented to allow new users to be generated upon their first signup.

**Acceptance Criteria**
- All mock user IDs are successfully replaced with real user IDs.
- The new user creation service is implemented and functional.
- The application correctly handles user IDs during signup and subsequent interactions.

### PR DESCRIPTION
**Pull request by kinskiwu**
## Description
Frontend:
- Replaced all mock user data with real user data.
- Added a default 5-minute stale time for the global query client to reduce unnecessary fetching.
- Removed one of the duplicated sign-out buttons.
- Added validation for name input.
- Implemented snackbarNotifications in auth-related components.
- Created two new React Query hooks for fetching and creating user profiles.
- Enhanced auth components to use the new RH hooks

Backend:
- Implemented a new user route with functionalities to fetch user profile and create user profiles.
- Updated `FirebaseAuthMiddleware` by adding a method to find users by UID and improved how user data is passed down to controllers.
- Updated the `UserModel` to better integrate with Firebase auth and the future user profile management feature.
- Updated logging in various controllers and services.

Fixes #289 

## Type of change
- [X] ✨  New feature (non-breaking change which adds functionality)

## How Can this be tested? Testing Plan to review this PR
- [X] Where in the app does this change take place?
Auth
- [X] What are the steps to test this change?
1. At the root directory, run `npm run dev`.

**Sign Up**
2. Click `Sign In`, then `Sign Up`, and test Google sign-up and email-password sign-up (GitHub sign-up is not fully implemented yet).

**Sign Out**
3. Click `Sign Out`.

**Sign In**
4. Click `Sign In`, and use the information used during sign-up to test Google sign-in and email-password sign-in.

- [X] What is the expected output?
5. You should be able to see your username displayed on the side navigation upon sign-in/sign-up.
image

## Checklist:
- [X] I have performed a self-review of my own code.
- [X] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [X] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [X] I have reviewed the [**Security Best Practices Document**](./docs/SECURITY_BEST_PRACTICES.md).
- [X] I have followed the [**Naming Conventions Guide**](./docs/NAMING_CONVENTIONS_GUIDE.md) for my changes.
- [X] I have ensured that my pull request title is descriptive.

**Additional Information**
Related documentation updates will be handled in #276

### CODE REVIEW

File: server/controllers/genAiController.ts
Comment (LucieSeidler) [inline server/controllers/genAiController.ts:1]: I'm not sure I understand why this was deleted

**Review by LucieSeidler — COMMENTED** [review]
The user information in the top left still reads test-user

File: server/controllers/genAiController.ts
Comment (kinskiwu) [reply server/controllers/genAiController.ts:1]: Good call. I shouldnt have removed it.

**Review by LucieSeidler — APPROVED** [review]
Comment for testers : make sure you sign up before you sign in

**Comment by kinskiwu** [comment]
@LucieSeidler and I identified that when a user signs up, the path correctly switches to /generate, but sometimes the UI continues to display the auth page. Since this issue doesn't impact the functionality of this feature I created a bug ticket for it #313

**Comment by relandboyle** [comment]
> @LucieSeidler and I identified that when a user signs up, the path correctly switches to /generate, but sometimes the UI continues to display the auth page. Since this issue doesn't impact the functionality of this feature I created a bug ticket for it #313

@kinskiwu Are you able to confirm that this bug is not introduced by this PR? I have not experienced the described behavior or heard of anyone else experiencing it on another branch.

**Review by relandboyle — CHANGES_REQUESTED** [review]
> @LucieSeidler and I identified that when a user signs up, the path correctly switches to /generate, but sometimes the UI continues to display the auth page. Since this issue doesn't impact the functionality of this feature I created a bug ticket for it #313

Are we accounting for valid cookies remaining on the browser as we are testing behaviors with multiple users? I ask because the behavior you're describing sounds like what I see happening when log in, leave the site without logging out, and then return to the site. I'm seeing the static layout for logged-in users, but I'm not being redirected to the correct route / view, and I'm seeing numerous logs from the User Service showing "User not found". I'm also seeing the authentication middleware logging my correct UID, so there's clearly some logical disconnect here.

Login / logout, authentication, and router redirects appear to be behaving normally in the dev branch, so I think it's fair to say there is a regression in this branch and it should not be merged as-is.

Please involve @colinvandergraaf and/or myself so we can identify and correct the regression. 

2024-07-11 11:51:08:491 [DEBUG] [UserService] services\userService.ts 18:14 Object.findUserByUid
 Finding user with UID: (my correct Google UID displays here)       

2024-07-11 11:51:08:563 [WARN] [UserService] services\userService.ts 23:16 Object.findUserByUid
 User not found.

2024-07-11 11:51:08:564 [DEBUG] [User Controller] controllers\userController.ts 28:14 fetchUserProfile
 Fetching user profile for userId: undefineddd

2024-07-11 11:51:08:564 [DEBUG] [UserService] services\userService.ts 45:14 Object.findUserById
 Finding user with userId: undefined

2024-07-11 11:51:08:574 [WARN] [UserService] services\userService.ts 23:16 Object.findUserByUid
 User not found.

2024-07-11 11:51:08:575 [DEBUG] [Input Controller] controllers\userInputController.ts 118:14 fetchInputHistory        
 Fetching input history for user: undefined

2024-07-11 11:51:08:575 [DEBUG] [UserInput Service] services\userInputService.ts 60:14 Object.fetchInputHistory       
 Fetching input history { userId: undefined, page: 1, limit: 5, skip: 0 }

2024-07-11 11:51:08:647 [WARN] [UserService] services\userService.ts 50:16 Object.findUserById
 User not found.

2024-07-11 11:51:08:659 [DEBUG] [Input Controller] controllers\userInputController.ts 130:14 fetchInputHistory        
 Fetched input history: []

**Comment by kinskiwu** [comment]
@relandboyle This error appears when we use an account previously signed up with linkta before this feature was implemented. In this new feature, a user profile is created after signing up (which we didn't have previously thus it is showing a user profile not found error). To test this feature, please first sign up with your account.

**Comment by relandboyle** [comment]
@kinskiwu I deleted my accounts from Firebase and I'm seeing the same or similar behavior.

Can we schedule a Zoom discussion / review for this? Continued via Slack...

**Comment by kinskiwu** [comment]
Converting this PR to draft to fix some issues

**Review by relandboyle — APPROVED** [review]
Nice.

**Comment by kinskiwu** [comment]
Just fyi in case any of you are following this PR:
- @relandboyle @ejwoods @chrisr0892 and I had a live pair programming session today and we were able to identify and fix a couple of issues causing the errors.

- Changes I made to fix the residual SignUp Page issue: I added a userEffect to redirect a user to `/generate’  when a user is authenticated
- Changes I made to ensure a user signing in through oauth gets a new userprofile generated:  updated handleGoogleAuthCick & handleGithubAuthClick to create a new userProfile if it is nonexistent upon sign-in

**Pull request merged**